ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java

```diff
@@ -306,22 +314,72 @@ private void recover(MessageStoreRecovery recovery) throws StoreException, IOExc
 
   /**
    * Sets the end offset of all log segments.
+   * <p/>
+   * Assumed to run only on startup.
    * @throws IOException if an end offset could not be set due to I/O error.
    */
   private void setEndOffsets() throws IOException {
-    for (Offset segmentStartOffset : indexes.keySet()) {
-      Offset nextIndexSegmentStartOffset = indexes.higherKey(segmentStartOffset);
+    for (Offset segmentStartOffset : indexesRef.get().keySet()) {
+      Offset nextIndexSegmentStartOffset = indexesRef.get().higherKey(segmentStartOffset);
       if (nextIndexSegmentStartOffset == null || !segmentStartOffset.getName()
           .equals(nextIndexSegmentStartOffset.getName())) {
         // this is the last index segment for the log segment it refers to.
-        IndexSegment indexSegment = indexes.get(segmentStartOffset);
+        IndexSegment indexSegment = indexesRef.get().get(segmentStartOffset);
         LogSegment logSegment = log.getSegment(segmentStartOffset.getName());
         logSegment.setEndOffset(indexSegment.getEndOffset().getOffset());
       }
     }
   }
 
   /**
+   * @return the map of {@link Offset} to {@link IndexSegment} instances.
+   */
+  ConcurrentSkipListMap<Offset, IndexSegment> getIndexSegments() {
+    return indexesRef.get();
+  }
+
+  /**
+   * Atomically adds {@code filesToAdd} to and removes {@code segmentsToRemove} from the map of {@link Offset} to
+   * {@link IndexSegment} instances.
+   * @param filesToAdd the backing files of the {@link IndexSegment} instances to add.
+   * @param segmentsToRemove the start {@link Offset} of {@link IndexSegment} instances to remove.
+   * @throws IllegalArgumentException if any {@link IndexSegment} that needs to be added has an offset higher than that
+   * in the journal.
+   * @throws StoreException if an {@link IndexSegment} instance cannot be created for a provided segment to add.
+   */
+  void changeIndexSegments(List<File> filesToAdd, Set<Offset> segmentsToRemove) throws StoreException {
+    TreeMap<Offset, IndexSegment> segmentsToAdd = new TreeMap<>();
+    for (File indexSegmentFile : filesToAdd) {
+      IndexSegment indexSegment = new IndexSegment(indexSegmentFile, true, factory, config, metrics, journal);
+      segmentsToAdd.put(indexSegment.getStartOffset(), indexSegment);
+    }
+
+    indexesChangeLock.lock();
+    try {
+      if (segmentsToAdd.size() > 0) {
+        Offset lastSegmentToAddEndOffset = segmentsToAdd.lastEntry().getValue().getEndOffset();
```

Comment:
I'm not adding from the last - just checking whether the end offset of the last entry >= journal's first offset since this is a tree map and it is sorted. Anyway, this has gone away in the refactoring.